### DISCUSSION TITLE
New Apache POI 5.2.2 not working

### DISCUSSION THREAD

**Discussion by apple162883** [nextgenhealthcare/connect#5358] [Q&A] [answered]
I was using Apache POI 3.9, POI-OOXML 3.8 and POI-OOXML-SCHEMAS-3.7  version to read .xlsx files from Mirth and it was working perfectly. Now I updated the JARs to newer versions of Apache POI 5.2.2, POI-OOXML 5.2.2 and POI-OOXML-SCHEMAS-4.1.2 and changed the import process also according to the new versions.
var workBook = new org.apache.poi.xssf.usermodel.XSSFWorkbook(inputstream);
var sheet = Packages.org.apache.poi.xssf.usermodel.XSSFSheet;

But now I am getting below errors:
1. java.lang.NoClassDefFoundError: org/apache/commons-io/ ----- I solved this by placing commons-io jar in custom-lib.

Now the error is 
java.lang.NoClassDefFoundError: org/apache/logging/log4j/LogManager
	at org.apache.poi.ooxml.POIXMLDocumentPart.<clinit>(POIXMLDocumentPart.java:56)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)

Has there been some changes in newer Apache version? If not do I have to place each additional jar files now individual in the library?
If anyone has ideas please provide them.

**Comment by ab-20-m** [comment, marked as answer]
Apache POI 5.2.2 has a dependency on log4j2 [amongst other things.](https://mvnrepository.com/artifact/org.apache.poi/poi/5.2.2).

**Reply by tonygermano** [reply]
Mirth 4.1.0 ships with log4j2

**Reply by apple162883** [reply]
For Mirth Connect they are using commons-io-2.6 jar as default. Apache POI 5.2 uses commons-io-2.11.0. When I try to run the default commons-io-2.6 is getting picked up. This is throwing function not found byteArray() error.
Is there any way to override this default jar with our version in Mirth?

**Reply by jonbartels** [reply]
You should thumbs-up https://github.com/nextgenhealthcare/connect/issues/4897 . @is-simon is asking for updates that would help you. That issue is slightly different in that a plugin is involved but the fundamental issue is the same. 

What specific error is being thrown?

The "safe" option is to load your newer POI JAR and all of its dependencies to a resource directory. The use the `IsolatedClassLoader` in your JS transformers to load the newer POI and its dependencies. This will work but is quite tedious to set up. You can read about it at https://github.com/nextgenhealthcare/connect/issues/3718 and at forums.mirthproject.io.

It is *sometimes* possible to upgrade libraries in MC by simply deleting the old one from `server-lib` and then loading the newer one to `server-lib`. I would generally expect Apache tooling to follow semver.org norms so 2.11 would be backwards compatible with 2.6 (but 3.1 would NOT be backward compatible with 2.x) however it is not guaranteed. 

Loading the 2.11 JAR to custom-lib or to server-launcher-lib will probably not work. Java classpath loads JARs where the first one loaded has precedence. It might be worth trying in case you get lucky, but expect it to fail.

**Reply by apple162883** [reply]
@jonbartels Thank you for your response. I used the IsolatedClassLoader to bypass the server-lib. It is working fine now. Here is a snippet of the code.
**_Previous version:_**
var workBook = new org.apache.poi.xssf.usermodel.XSSFWorkbook(inputstream);
var row 	=  org.apache.poi.ss.usermodel.Row;
var sheet = Packages.org.apache.poi.xssf.usermodel.XSSFSheet;

**_With IsolatedClassLoader_** 
 var iso = Packages(contextFactory.getIsolatedClassLoader());
var workBook = new iso.org.apache.poi.xssf.usermodel.XSSFWorkbook(inputstream);
var row = iso.org.apache.poi.ss.usermodel.Row;
var sheet = Packages.iso.org.apache.poi.xssf.usermodel.XSSFSheet;

**Reply by jmcook74** [reply]
How did you determine that Mirth was running commons-oi-2.6? I am running into the same byteArray() error

"For Mirth Connect they are using commons-io-2.6 jar as default. Apache POI 5.2 uses commons-io-2.11.0. When I try to run the default commons-io-2.6 is getting picked up. This is throwing function not found byteArray() error.
Is there any way to override this default jar with our version in Mirth?"


EDIT:
I fixed this minutes after asking my question. I just moved my commons-io to appear earlier in the build path.
